graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,282 @@
+
+        byte[] bytes = trace.toByteArray();
+        tracingMap.put(KEY, Base64.getEncoder().encodeToString(bytes));
+
+        if (debuggingEnabled) {
+            tracingMap.put(DEBUG_KEY, trace.toString());
+        }
+
+        return CompletableFuture.completedFuture(new ExecutionResultImpl(executionResult.getData(), executionResult.getErrors(), tracingMap));
+    }
+
+    @Override
+    public InstrumentationContext<Object> beginFieldFetch(InstrumentationFieldFetchParameters parameters) {
+        FederatedTracingState state = parameters.getInstrumentationState();
+
+        long startNanos = System.nanoTime();
+        return whenCompleted((result, throwable) -> {
+            long endNanos = System.nanoTime();
+
+            DataFetchingEnvironment env = parameters.getEnvironment();
+            ExecutionStepInfo executionStepInfo = env.getExecutionStepInfo();
+
+            Node node = state.addNode(executionStepInfo, startNanos, endNanos);
+
+            if (throwable != null) {
+                node.addError(throwable);
+            }
+
+            if (result instanceof DataFetcherResult && ((DataFetcherResult) result).hasErrors()) {
+                DataFetcherResult theResult = (DataFetcherResult) result;
+                List<GraphQLError> errors = theResult.getErrors();
+                errors.forEach(node::addError);
+            }
+        });
+    }
+
+    /**
+     * Stores timing information and a map of nodes for field information.
+     */
+    private class FederatedTracingState implements InstrumentationState {
+        private final Instant startRequestTime;
+        private final long startRequestNanos;
+        private final LinkedHashMap<ExecutionPath, Node> nodeMap;
+
+        private FederatedTracingState() {
+            // record start time when creating instrumentation state for a request
+            startRequestTime = Instant.now();
+            startRequestNanos = System.nanoTime();
+
+            // Create node map with the root node
+            nodeMap = new LinkedHashMap<>();
+            nodeMap.put(ExecutionPath.rootPath(), new Node(ExecutionPath.rootPath()));
+        }
+
+        Reports.Trace toProto() {
+            return Reports.Trace.newBuilder()
+                    .setStartTime(getStartTimestamp())
+                    .setEndTime(getNowTimestamp())
+                    .setDurationNs(getDuration())
+                    .setRoot(buildNodeTree())
+                    .build();
+        }
+
+        /**
+         * Adds node to nodeMap and recursively ensures that all parent node exist.
+         */
+        Node addNode(ExecutionPath path, Node node) {
+            if (!nodeMap.containsKey(path)) {
+                nodeMap.put(path, node);
+            }
+
+            List<Object> pathParts = path.toList();
+            ExecutionPath parentPath = ExecutionPath.fromList(pathParts.subList(0, pathParts.size() - 1));
+
+            if (!nodeMap.containsKey(parentPath)) {
```

Comment:
Out of curiosity, since I'm only giving this a brief look atm: how can we get into a state like this? Would this be caused from a bug within the instrumentation pipeline around field execution in GraphQL Java?

Reply:
The reason for the backwards construction of node parents is to support indexes. The instrumentation hooks don't fire for `widgets.0`, just for `widgets.0.id`. I agree this isn't super clear though. I'll see what happens to this part of the code when I try dropping the wrapper class.